### ISSUE TITLE
Can not get params which application should receive during install

### ISSUE THREAD

**Issue by SergeyMagid** [BranchMetrics/Android-Deferred-Deep-Linking-SDK#97]
Hi,

I use version 1.5.9

Steps : 
1. I create share link.
2. Install app by this link.
3. In SplashScreen try to get params from deep link by use method getFirstReferringParams().
But there no params. 
https://github.com/BranchMetrics/Branch-Android-SDK#retrieve-install-install-only-parameters

Problem 
branch.getFirstReferringParams();
did not return expected params

**Comment by aaustin** [comment]
@SergeyMagid Were the parameters available in the deep link handler callback that you registered in the initSession call? Also, I'd recommend using _getLatestReferringParams_, which will return the set of parameters most recently clicked.

Get first _referring_ parameters are set only once, the first time the deep link is triggered on an 'install' event. You can get _first_ referring params to be set on 'open's too if you trigger isReferrable:true in the initSession call.

**Comment by aaustin** [comment]
@SergeyMagid Just wanted to check in that you got this all sorted out

**Comment by sojanpr** [comment]
@SergeyMagid  Are you still having the issue. Can you please make sure you are calling `branch.getFirstReferringParams();` a after initializing the branch session by calling ` branch.initSession()`

**Comment by rafalzawadzki** [comment]
I have the same problem, tried you suggestions but that does not work either.
Tried approaches:
1) Autodeeplinking (manifest meta + almost the same code as in you android example + getFirstReferringParams() + getLatestReferringParams())
2) Normal deep linking (manifest intent-filter + initSession() + getFirstReferringParams() + getLatestReferringParams())
3) Without changes in manifest (initSession() + getFirstReferringParams() + getLatestReferringParams())

In every case I can see debug info:
```
I/BranchSDK﹕ posting to https://api.branch.io/v1/install
I/BranchSDK﹕ Post value = {
    "hardware_id": "4a79a66068c3ccf9",
    "is_hardware_id_real": true,
    "app_version": "1.00",
    "carrier": "PLAY",
    "bluetooth": false,
    "bluetooth_version": "ble",
    "has_nfc": true,
    "has_telephone": true,
    "brand": "LGE",
    "model": "Nexus 5",
    "os": "Android",
    "os_version": 22,
    "screen_dpi": 480,
    "screen_height": 1776,
    "screen_width": 1080,
    "wifi": false,
    "is_referrable": 0,
    "update": 2,
    "debug": false,
    "google_advertising_id": "fd80bad8-ae6f-432f-b1cd-b55e489b5d9f",
    "lat_val": 1,
    "sdk": "android1.8.1",
    "retryNumber": 0,
    "branch_key": "key_live_mcnLaL1KLTW5vsJt8sO9Epnetymi1nd2"
    }
I/BranchSDK﹕ returned {"session_id":"157420067066222447","identity_id":"157277543118935585","device_fingerprint_id":"157277543014112026","browser_fingerprint_id":null,"link":"http://r.bitween.co/j/5C4myvNuMl","referring_identity":"simplelogin:27","identity":"simplelogin:27","data":"{\"$og_description\":\"Join Joe in My App - it's awesome\",\"$og_image_url\":\"https://s3-us-west-1.amazonaws.com/myapp/joes_pic.jpg\",\"$og_title\":\"Joe's My App Referral\",\"avatar\":\""
```
An then my logs that contain data returned by getFirstReferringParams() and getLatestReferringParams(), as you can see there is only "+clicked_branch_link, true" in last data:
```
D/BranchHelper﹕ Branch init complete.
D/BranchHelper﹕ Reading params.
I/last﹕ +clicked_branch_link, true
```

One more thing worth mentioning is in my manifest I have declared io.branch.sdk.TestMode="true", but Branch only uses live key and debug in dashboard also works only in Live mode.

Your help would be greatly appreciated. I would like to add referral links with welcoming screens in my app but for couple of hours I am unable to find proper solution.

**Comment by sojanpr** [comment]
@rafalzawadzki Thanks for the detailed report. We are looking in to the issue .Could you please update the version of Branch SDK that you are using

**Comment by rafalzawadzki** [comment]
@sojanpr The "avatar" field contained image in base64 string and in sake of post simplicity I cut it off - now I dont know if the structure was fine before.
I will try again without so much data in the json - now it striked me that this may be the problem!

**Comment by rafalzawadzki** [comment]
Ok, now it looks much better:

```
I/BranchSDK﹕ posting to https://api.branch.io/v1/install
I/BranchSDK﹕ Post value = {
    "hardware_id": "4a79a66068c3ccf9",
    "is_hardware_id_real": true,
    "app_version": "1.01",
    "carrier": "PLAY",
    "bluetooth": false,
    "bluetooth_version": "ble",
    "has_nfc": true,
    "has_telephone": true,
    "brand": "LGE",
    "model": "Nexus 5",
    "os": "Android",
    "os_version": 22,
    "screen_dpi": 480,
    "screen_height": 1776,
    "screen_width": 1080,
    "wifi": false,
    "is_referrable": 0,
    "update": 2,
    "debug": false,
    "google_advertising_id": "...",
    "lat_val": 1,
    "sdk": "android1.8.1",
    "retryNumber": 0,
    "branch_key": "..."
    }
I/BranchSDK﹕ returned {"session_id":"157747274057783099","identity_id":"157277543118935585","device_fingerprint_id":"157277543014112026","browser_fingerprint_id":null,"link":"http://r.bitween.co/j/5C4myvNuMl","referring_identity":"simplelogin:27","identity":"simplelogin:27","data":"{\"$og_description\":\"Join Joe in My App - it's awesome\",\"$og_image_url\":\"https://s3-us-west-1.amazonaws.com/myapp/joes_pic.jpg\",\"$og_title\":\"Joe's My App Referral\",\"nick\":\"master\",\"source\":\"android\",\"~channel\":\"Copy link\",\"~feature\":\"share\",\"~tags\":[\"version_213\"],\"~stage\":\"post_purchase\",\"~creation_source\":\"SDK\",\"+is_first_session\":false,\"+clicked_branch_link\":true}"}
D/BranchHelper﹕ Branch init complete.
D/BranchHelper﹕ Reading params.
I/last﹕ $og_description, Join Joe in My App - it's awesome
I/last﹕ $og_image_url, https://s3-us-west-1.amazonaws.com/myapp/joes_pic.jpg
I/last﹕ $og_title, Joe's My App Referral
I/last﹕ nick, master
I/last﹕ source, android
I/last﹕ ~channel, Copy link
I/last﹕ ~feature, share
I/last﹕ ~tags, ["version_213"]
I/last﹕ ~stage, post_purchase
I/last﹕ ~creation_source, SDK
I/last﹕ +is_first_session, false
I/last﹕ +clicked_branch_link, true
```
It looks like too much data breaks your internal processes, or something.
Now, I hae a question: how I can know this is fresh install data? As you can see I can get only info from getLatestReferringParams() method and it only returns "+is_first_session, false".

**Comment by sojanpr** [comment]
@rafalzawadzki Glad to see that you are  getting referrer params. Instead of adding large `avatar` param to the JSON, is it possible to add a URL or Identifier which points to the data.

Thank you very much  for mentioning the issue with the `BranchKey` and install data. We figured out a  bug on our side while initializing Branch from custom application class. This issue is already fixed and released with version v1.8.2. Could you please try to update  your sdks to 1.8.2 and try out.
 
Branch actually save  a `Hardware-ID` on the first install. So even if you uninstall and and install again Branch  will return "+is_first_session, false".  This is to prevent fake installation count. 

To test installtion you can set your app `TestMode`. In test mode Branch creates random  `Hardware-ID` and so you will get  "+is_first_session, true" each time you install.

You can set  app to test mode by setting following metadata in your manifest
```xml
<application>
.......
<meta-data android:name="io.branch.sdk.TestMode" android:value="false" />
....
</application>
```
Hope this will help you. Please let us know if you have any issues

**Comment by sojanpr** [comment]
@rafalzawadzki  Just wanted to check whether the fix helped to solve the issue.

**Comment by rafalzawadzki** [comment]
@sojanpr I believe this should be stated as "TestMode=true" instead of "testMode=false" and indeed the problem with the keys is solved - now Branch uses test keys correctly :)

But I still have the problem with "+is_first_session", cannot create random HardwareId and fake fresh install. What is the proper way to implement deeplinking? Are AndroidManifest tags "intent-filter" and "meta" necessary and if so, which one should I use? It seems that Branch initSession works well without these, so what are they needed for? Will be glad for clearing things up.

**Comment by aaustin** [comment]
@rafalzawadzki We found a bug with the library today where setDebug was not actually setting the debug. The only mechanism to enable debug was this:

```xml
<application>
.......
<meta-data android:name="io.branch.sdk.TestMode" android:value="false" />
....
</application>
```

However, then I found that this ^ would only work if you had a TestKey in the manifest as well.. I just made and merged https://github.com/BranchMetrics/Android-Deferred-Deep-Linking-SDK/pull/115 that deprecates setDebug an allows TestMode to work with the Live key as well.

We're about to deploy it to Maven shortly.

**note:** for testing, if the hardware_id is in this form '4a79a66068c3ccf9', that's a real one. If it looks like a UUID '284b3d7a-57f4-4930-a938-56e2bd46124c', then it's fake one. 

Really sorry for this headache.. not sure how we didn't catch this earlier.

**Comment by rafalzawadzki** [comment]
Ok silly me, I found the solution. 
The problem was no with Branch itself but with Lollipop and multiple user accounts. Starting with Lollipop, sometimes apps are not unistalled for all the users and admin has to enter settings to uninstall the app completely (more on this here: http://stackoverflow.com/questions/28853250/you-cannot-install-this-app-because-another-user-has-already-installed-an-incom) Thats why Branch was thinking (correctly) that the app was installed before.
By the way, you could mention it here: https://dev.branch.io/recipes/testing_your_integration/android/ as it is very possible someone will stumble on this in the future - and the cause is not as clear as the solution :)

**Comment by aaustin** [comment]
Ahh - I got it. Thanks for the note! We'll update the docs.

Closing this ticket for now.

**Comment by Pre8y** [comment]
I have added 
<meta-data
            android:name="io.branch.sdk.BranchKey"
            android:value="@string/branch_io_key" />

        <meta-data android:name="io.branch.sdk.BranchKey.test" android:value="@string/branch_io_test_key" />
        <meta-data android:name="io.branch.sdk.TestMode" android:value="true" />

I have uninstalled the app from all users and re-installing it, still I am getting +is_first_session false.

**Comment by Sarkar** [comment]
@Pre8y we have the conversation at support, and it looks like that took care of it, as you're seeing +is_first_session as true. What was your fix? Were you uninstalling and re-installing the .apk?

**Comment by Pre8y** [comment]
@Sarkar +is_first_session is coming false always, and +is_branch_clicked is also coming false always. even after enabling testMode in Manifest and adding branch test key in Manifest.

**Comment by rafalzawadzki** [comment]
@Pre8y hey, are you cleaning your browser cookies etc before clicking branch link?

**Comment by Pre8y** [comment]
yes tried that as well.

**Comment by aaustin** [comment]
Can you check if the Branch key you created the link with matches the key you're testing with in the manifest? We don't allow deep linking across keys. You can check the link by appending ?debug=true to it and pasting it into the browser on your desktop.

**Comment by novan** [comment]
I have a weird issue here. I can't get the referring params on post install. But I can get it on running the app for the second time. What do I miss?

**Comment by aaustin** [comment]
Hmm - it would seem that Branch might not be initializing post install.
- Can you repeat it 100% of the time? I want to check that it's not an IP mismatch issue. We've heard of offices that randomly assign IPs between each request that can make testing pretty difficult.
- Can you add `<meta-data android:name="io.branch.sdk.TestMode" android:value="true" />` above your Branch key in the manifest to log Branch requests? This will tell you if Branch is initializing post install.
- I want to confirm you're calling getAutoInstance in the Application class: https://dev.branch.io/getting-started/sdk-integration-guide/guide/android/#enable-auto-session-management

**Comment by saeidfarivar** [comment]
@aaustin, I have the same issue. first launch doesn't initialize properly (don't have the session params). but second launch has it all right.
I can reproduce this 100%
I have getAutoInstance in my Application class. 
I am using version 1.14.1 on android.

**Comment by aaustin** [comment]
Hey @saeidfarivar - quick question:
- when you enable debug mode, do you see a network call to /v1/install or /v1/open on first launch? Or is there no network call at all.
- what make/model phone are you testing with?
 
We haven't been able to reproduce well. Is it possible that you create a simple testbed app that has the same behavior or we could get access to your project in a private repo so that we could play around with it?

**Comment by saeidfarivar** [comment]
Yes, I do see a call to  /v1/install in the logs.

Maybe my test is wrong. 
How I test this:
install the app,
clear App data from settings,
click on Branch link,
result => my logs says this is not the first session? (+is_first_session is false)

How should I test my integration for the deferred installs?

**Comment by sojanpr** [comment]
@saeidfarivar Can you please make sure you have enabled `TestMode` for doing the above test.
(Check that you have `<meta-data android:name="io.branch.sdk.TestMode" android:value="true" /> `in your manifest) .
 In `Live` more Branch can detect between `first install  ` and ` re-install` even if you delete the application data

**Comment by saeidfarivar** [comment]
@sojanpr yes, I have <meta-data android:name="io.branch.sdk.TestMode" android:value="true" /> in my manifest. still the branch sdk doesn't give me the right session params...

**Comment by sojanpr** [comment]
@saeidfarivar  Sorry forgot to mention. Since you are creating instance of Branch in your application class could you please  try `getTestInstance` instead of `getInstance`

```public class YourApp extends Application {
    @Override
    public void onCreate() {
        super.onCreate();      
        Branch.getTestInstance(this);        
    }
}```

**Comment by saeidfarivar** [comment]
@sojanpr I did that too. no change.
Our office is 5 minutes away from your Palo Alto, CA office. I cannot share code but if you guys can come here I can show you the code...

**Comment by sojanpr** [comment]
@saeidfarivar Sorry for this issue. We are able to reproduce this issue on our side with the install. Working on a way around for this. Update you asap

**Comment by sojanpr** [comment]
@saeidfarivar The latest version of our SDK is 2.2.0. The above case is working fine with the latest version. Could you please  try it and update?

**Comment by saeidfarivar** [comment]
@sojanpr with 2.2.0 it's not calling onInitFinished at all!!! 
And there is no debug logs from Branch (e.g. post /v1/install)...

**Comment by saeidfarivar** [comment]
@sojanpr I use your jar file from https://github.com/BranchMetrics/android-branch-deep-linking
am I missing something?

**Comment by sojanpr** [comment]
@saeidfarivar could you please try the gradle dependency instead of the  Jar
Just add compile 'io.branch.sdk.android:library:2.+' to the dependencies section of your build.gradle file.

**Comment by saeidfarivar** [comment]
@sojanpr, I built your repository and took the .aar file. still not working. Our project don't use gradle.

**Comment by sojanpr** [comment]
@saeidfarivar really sorry about that. We are working on it. Could you please help us by sending your logs? It would greatly help us if we can see the Branch logs for `v1/install ` api request and response in this case

**Comment by saeidfarivar** [comment]
@sojanpr , I don't see that in the logs... in fact it seems that Branch doesn't print any logs anymore...
using 1.14.3 I used to see a lot of logs after enabling test mode.

**Comment by sojanpr** [comment]
@saeidfarivar could you please make sure you have 
`<meta-data android:name="io.branch.sdk.TestMode" android:value="true" /> ` in your manifest and  you are using `getAutoTestInstance()` . Logs will be printed only in test mode

**Comment by saeidfarivar** [comment]
@sojanpr, Oops I thought you said 

> 


@saeidfarivar Sorry forgot to mention. Since you are creating instance of Branch in your application class could you please try getTestInstance instead of getInstance

public class YourApp extends Application { @Override public void onCreate() { super.onCreate(); Branch.getTestInstance(this); } }

**Comment by sojanpr** [comment]
@saeidfarivar Inorder to help you better and faster could you please send an email to sojan@branchmetrics.io the logs.  Lets us take it from there

**Comment by saeidfarivar** [comment]
after setting <meta-data android:name="io.branch.sdk.TestMode" android:value="true" /> and using getAutoTestInstance() and updating to sdk 2.2.0. everything seems to be working for me. thanks @sojanpr

**Comment by zjxuzhj** [comment]
@rafalzawadzki  Really is very grateful that you solved my problem

**Comment by EvangelosG** [comment]
👍
